### ISSUE TITLE
don't run `yarn check --integrity` on every rails command

### ISSUE THREAD

**Issue by doits** [rails/webpacker#1135]
`yarn check --integrity` is always run. Even if I do a `rails g migration ...`, which obviously has nothing to do with yarn.

I benchmarked it and noticed it adds about 400ms delay to such commands in my case, which is huge imo. It should be changed that it only runs when it makes sense to check for it, like when starting up the web server or the webpack-dev-server.

Since I am not familiar with the rails/railtie internals, I don't know if this is possible or how to implement such checks at the right place.

Is this possible? What do you think?

**Comment by kriansa** [comment]
You can disable that in `development.rb`. 

Just set `config.webpacker.check_yarn_integrity = false`

**Comment by doits** [comment]
@kriansa yes, I know this, but this disables it completely. It often saved me already when running `bin/webpacker` or `rails server` with outdated deps, so in general it makes sense to have it enabled on development. But not when I run `rails g migration`.

**Comment by doits** [comment]
Why was this closed? Will/can not be fixed?

**Comment by ansonhoyt** [comment]
This check also breaks adding webpacker to an existing project if `yarn.lock` exists. Running `rails webpacker:install` loads rails, so this initializer tries to fetch :check_yarn_integrity, which loads `config/webpacker.yml`. However, this doesn't exist until after we run this rake task.

I had a chuckle at the error message suggesting I run `rails webpacker:install`:
```
$ bundle exec rails webpacker:install
RAILS_ENV=development environment is not defined in config/webpacker.yml, falling back to production environment
rails aborted!
Webpacker configuration file not found /Users/ahoyt/code/tiu/pd-tracker/config/webpacker.yml. Please run rails webpacker:install Error: No such file or directory @ rb_sysopen - /Users/ahoyt/code/tiu/pd-tracker/config/webpacker.yml
```

We could update the [installation instructions](https://github.com/rails/webpacker#installation) to:
```
rm yarn.lock
bundle
bundle exec rails webpacker:install
```

...but that seems hacky. I like the idea of this initializer being smarter. I haven't learned enough about it to suggest how.

**Comment by mcmire** [comment]
I would love to see this revisited. If anything, I feel like `yarn.lock` should be checked on running `rails assets:precompile` (and, as mentioned, on running `bin/webpack` or `bin/webpack-dev-server`). Otherwise, it doesn't really matter, since Webpack is run in a separate process.

**Comment by aliraxa-max** [comment]
How I can resolve this problem,
=> Booting Puma
=> Rails 6.0.0.rc2 application starting in development 
=> Run `rails server --help` for more startup options
error Couldn't find an integrity file                                                                                                                
error Found 1 errors.                                                                                                                                


========================================
  Your Yarn packages are out of date!
  Please run `yarn install --check-files` to update.
========================================


To disable this check, please change `check_yarn_integrity`
to `false` in your webpacker config file (config/webpacker.yml).


yarn check v1.17.3
info Visit https://yarnpkg.com/en/docs/cli/check for documentation about this command.


Exiting

**Comment by chiptus** [comment]
Run `rm - rf node_modules && yarn`

**Comment by korny** [comment]
This is quite annoying. The check now also runs `yarn check --verify-tree`, which takes a whooping 2.7s on my system, effectively doubling environment loading time on every rails command:

```
# check_yarn_integrity: false
$ time rails runner ''

real	0m3.613s
user	0m2.496s
sys	0m0.790s

# check_yarn_integrity: true
$ time rails runner ''

real	0m6.773s
user	0m5.696s
sys	0m1.575s
```

Like the OP, I would like to keep the check for `rails s`, but I don't see why it needs to be run for everything else.

**Comment by Grunkhead** [comment]
I fixed this by doing the following steps;

I had to create a symlink because the node binary was not found.
sudo ln -s /usr/bin/nodejs /usr/bin/node

Upgrading NodeJS to 13.* (probably swapped the binary path here, symlink might have been overkill).

Updating Yarn, [https://yarnpkg.com/en/package/yarn](url)
Refreshing the shell instance.

**Comment by DouglasUrner** [comment]
> I fixed this by doing the following steps;
> 
> I had to create a symlink because the node binary was not found.
> sudo ln -s /usr/bin/nodejs /usr/bin/node
> 
> Upgrading NodeJS to 13.* (probably swapped the binary path here, symlink might have been overkill).
> 
> Updating Yarn, [https://yarnpkg.com/en/package/yarn](url)
> Refreshing the shell instance.

Hmm, I'm seeing it with Rails 6 and Node 13.6...

**Comment by yasirlateef** [comment]
I was able to fix it with the following:

1. Upgrade Yarn: `curl --compressed -o- -L https://yarnpkg.com/install.sh | bash`
2. remove node_modules and yarn.lock
3. yarn install

**Comment by david-forer** [comment]
https://github.com/rails/webpacker/issues/1135#issuecomment-607972113 this worked for me. Thank You @yasirlateef

**Comment by peterwillcn** [comment]
config.webpacker.check_yarn_integrity = false

**Comment by kurko** [comment]
An environment variable would be the best approach IMO. That way you can have it true by default when running `rails server` or similar, and false otherwise.

**Comment by ttilberg** [comment]
This issue got pinged from a mention above, but it's worth pointing out that this might be able to be closed:

https://github.com/rails/webpacker/pull/2518

It looks like you can webpacker to help this out (and this can probably be closed).

**Comment by rossta** [comment]
@doits Can we close this issue? Fixed in #2518.
